### PR TITLE
[Promotion] Coupon prefix and suffix

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

Form: 
Zrzut ekranu 2019-07-1 o 12 50 41

Generated coupons:
Zrzut ekranu 2019-07-1 o 12 59 35

### CODE REVIEW

**Comment by CoderMaggie** [comment]
Would be wonderful if we could add tooltips explaining what are these.
Moreover, I would improve the UX, by making the code length field shorter and the prefix/suffix fields with some placeholders `I.e. 2019SALE_`

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature [hunk 1]
```diff
@@ -0,0 +1,46 @@
+@managing_promotion_coupons
+Feature: Generating a new coupons with prefix and suffi
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]:
```suggestion
Feature: Generating new coupons with prefix and suffix
```

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    In order to quickly create specific number of tailored coupons for my promotions
+    As an Administrator
+    I want to be able to generate coupons for promotion with prefix and suffix
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]:
```suggestion
    I want to be able to generate promotion coupons with prefix and suffix
```

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    @ui
+    Scenario: Generating new coupons with prefix
+        Given I want to generate a new coupons for this promotion
+        When I specify its amount as 5
+        And I specify its prefix as "CHRISTMAS_"
+        And I specify its code length as 6
+        And I generate it
+        Then I should be notified that it has been successfully generated
+        And there should be 5 coupon related to this promotion
+        And all of the coupons codes should be prefixed with "CHRISTMAS_"
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]:
```suggestion
        And all of the coupon codes should be prefixed with "CHRISTMAS_"
```

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    @ui
+    Scenario: Generating new coupons with prefix
+        Given I want to generate a new coupons for this promotion
+        When I specify its amount as 5
+        And I specify its prefix as "CHRISTMAS_"
+        And I specify its code length as 6
+        And I generate it
+        Then I should be notified that it has been successfully generated
+        And there should be 5 coupon related to this promotion
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]:
```suggestion
        And there should be 5 coupons related to this promotion
```

**Comment by Zales0123** [comment]
@CoderMaggie agreed 👍 However, as it's a UI/UX job, I would leave it as it is in this PR (to not make it longer than needed) and let somebody more experienced with front-end (probably @kulczy 😄) do the job 🚀

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And there is a promotion "Christmas sale"
+        And it is coupon based promotion
+        And I am logged in as an administrator
+
+    @ui
+    Scenario: Generating new coupons with prefix
+        Given I want to generate a new coupons for this promotion
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]:
```suggestion
        Given I want to generate new coupons for this promotion
```

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        When I specify its amount as 5
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]:
```suggestion
        When I choose the amount of 5 coupons to be generated
```

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        And I specify its prefix as "CHRISTMAS_"
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]:
```suggestion
        And I specify their prefix as "CHRISTMAS_"
```

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        And I specify its code length as 6
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]:
```suggestion
        And I specify their code length as 6
```

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        And I generate it
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]:
```suggestion
        And I generate these coupons
```

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        Then I should be notified that it has been successfully generated
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]:
```suggestion
        Then I should be notified that they have been successfully generated
```

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    @ui
+    Scenario: Generating new coupons with suffix
+        Given I want to generate a new coupons for this promotion
+        When I specify its amount as 5
+        And I specify its suffix as "_CHRISTMAS"
+        And I specify its code length as 6
+        And I generate it
+        Then I should be notified that it has been successfully generated
+        And there should be 5 coupon related to this promotion
+        And all of the coupons codes should be suffixed with "_CHRISTMAS"
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]: same as above, this whole scenario

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        Given I want to generate a new coupons for this promotion
+        When I specify its amount as 5
+        And I specify its prefix as "CHRISTMAS_"
+        And I specify its suffix as "_SALE"
+        And I specify its code length as 6
+        And I generate it
+        Then I should be notified that it has been successfully generated
+        And there should be 5 coupon related to this promotion
+        And all of the coupons codes should be prefixed with "CHRISTMAS_"
+        And all of the coupons codes should be suffixed with "_SALE"
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]: same as above, this whole scenario

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature (hunk 1 above)
Comment (Tomanhez) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]:
```suggestion
Feature: Generating new coupons with prefix and suffi
```

plural

File: src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionCouponsContext.php [hunk 13]
```diff
@@ -248,6 +264,26 @@ public function thereShouldBeCouponRelatedTo($number, PromotionInterface $promot
...
+    {
+        foreach ($this->indexPage->getCouponCodes() as $couponCode) {
+            Assert::startsWith($couponCode, $prefix);
+        }
+    }
+
+    /**
+     * @Then all of the coupons codes should be suffixed with :prefix
+     */
+    public function allOfTheCouponsShouldBeSuffixedWith(string $suffix): void
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionCouponsContext.php]:
```suggestion
    public function allOfTheCouponsCodesShouldBeSuffixedWith(string $suffix): void
```

File: features/promotion/managing_coupons/generating_coupon.feature [hunk 14]
```diff
@@ -12,21 +12,21 @@ Feature: Generating a new coupons
...
         And I generate it
-        Then I should be notified that it has been successfully generated
+        Then I should be notified that they have been successfully generated
         And there should be 5 coupon related to this promotion
 
     @ui
     Scenario: Generating a new coupons without expiration date
-        Given I want to generate a new coupons for this promotion
+        Given I want to generate new coupons for this promotion
         When I specify its amount as 5
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon.feature]:
```suggestion
        When I choose the amount of 5 coupons to be generated
```

Comment (CoderMaggie) [reply features/promotion/managing_coupons/generating_coupon.feature]: Shall we make this change everywhere we use this step?

File: features/promotion/managing_coupons/generating_coupon.feature [hunk 15]
```diff
@@ -12,21 +12,21 @@ Feature: Generating a new coupons
...
-        And I specify its code length as 6
+        And I specify their code length as 6
         And I limit generated coupons usage to 25 times
         And I generate it
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon.feature]:
```suggestion
        And I generate these coupons
```
and this one also?

File: features/promotion/managing_coupons/generating_coupon.feature [hunk 16]
```diff
@@ -12,21 +12,21 @@ Feature: Generating a new coupons
...
-        Then I should be notified that it has been successfully generated
+        Then I should be notified that they have been successfully generated
         And there should be 5 coupon related to this promotion
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon.feature]:
```suggestion
        And there should be 5 coupons related to this promotion
```

File: features/promotion/managing_coupons/generating_coupon.feature [hunk 17]
```diff
@@ -12,21 +12,21 @@ Feature: Generating a new coupons
 
     @ui
     Scenario: Generating a new coupons
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon.feature]:
```suggestion
    Scenario: Generating new coupons
```

File: src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionCouponsContext.php [hunk 18]
```diff
@@ -239,7 +257,7 @@ public function iDeleteThem(): void
     }
 
     /**
-     * @Then /^there should be (\d+) coupon related to (this promotion)$/
+     * @Then /^there should be (\d+) coupon(?:|s) related to (this promotion)$/
```
Comment (CoderMaggie) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionCouponsContext.php]:
```suggestion
     * @Then /^there should be 1 coupon related to (this promotion)$/
     * @Then /^there should be (\d+) coupons related to (this promotion)$/
```

I know it won't work probably, but do you see what I'd like to achieve? :D
Forbid the `Then there should be 5 coupon...` usage in scenarios :D

File: features/promotion/managing_coupons/coupon_generate_validation.feature [hunk 19]
```diff
@@ -12,19 +12,19 @@ Feature: Coupon generate instruction validation
 
     @ui
     Scenario: Trying to generate a new coupons without specifying its amount
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/coupon_generate_validation.feature]:
```suggestion
    Scenario: Trying to generate new coupons without specifying their amount
```

File: features/promotion/managing_coupons/coupon_generate_validation.feature [hunk 20]
```diff
@@ -34,8 +34,8 @@ Feature: Coupon generate instruction validation
 
     @ui
     Scenario: Trying to generate a new coupons with amount and code length impossible to generate
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/coupon_generate_validation.feature]:
```suggestion
    Scenario: Trying to generate new coupons with amount and code length impossible to generate
```

File: features/promotion/managing_coupons/generating_coupon.feature [hunk 21]
```diff
@@ -12,21 +12,21 @@ Feature: Generating a new coupons
...
+        And I specify their code length as 6
         And I limit generated coupons usage to 25 times
         And I make generated coupons valid until "26.03.2017"
         And I generate it
-        Then I should be notified that it has been successfully generated
+        Then I should be notified that they have been successfully generated
         And there should be 5 coupon related to this promotion
 
     @ui
     Scenario: Generating a new coupons without expiration date
```
Comment (CoderMaggie) [inline features/promotion/managing_coupons/generating_coupon.feature]:
```suggestion
    Scenario: Generating new coupons without expiration date
```

File: src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionCouponsContext.php [hunk 22]
```diff
@@ -248,6 +266,26 @@ public function thereShouldBeCouponRelatedTo($number, PromotionInterface $promot
...
+     */
+    public function allOfTheCouponCodesShouldBePrefixedWith(string $prefix): void
+    {
+        foreach ($this->indexPage->getCouponCodes() as $couponCode) {
+            Assert::startsWith($couponCode, $prefix);
+        }
+    }
+
+    /**
+     * @Then all of the coupon codes should be suffixed with :prefix
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionCouponsContext.php]:
```suggestion
     * @Then all of the coupon codes should be suffixed with :suffix
```

File: src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionCouponsContext.php [hunk 23]
```diff
@@ -372,9 +410,9 @@ public function couponShouldNotExistInTheRegistry(PromotionCouponInterface $coup
     }
 
     /**
-     * @Then I should be notified that it has been successfully generated
+     * @Then I should be notified that they have been successfully generated
      */
-    public function iShouldBeNotifiedThatItHasBeenSuccessfullyGenerated()
+    public function iShouldBeNotifiedThatTheyHaveBeenSuccessfullyGenerated()
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionCouponsContext.php]: Missing return type

File: features/promotion/managing_coupons/coupon_generate_validation.feature [hunk 24]
```diff
@@ -11,34 +11,34 @@ Feature: Coupon generate instruction validation
...
 
     @ui
-    Scenario: Trying to generate a new coupons with amount and code length impossible to generate
-        Given I want to generate a new coupons for this promotion
-        When I specify its code length as 1
-        And I specify its amount as 20
+    Scenario: Trying to generate new coupons with amount and code length impossible to generate
+        When I want to generate new coupons for this promotion
+        And I specify their code length as 1
+        When I choose the amount of 20 coupons to be generated
```
Comment (GSadee) [inline features/promotion/managing_coupons/coupon_generate_validation.feature]:
```suggestion
        And I choose the amount of 20 coupons to be generated
```

File: features/promotion/managing_coupons/coupon_generate_validation.feature [hunk 25]
```diff
@@ -11,34 +11,34 @@ Feature: Coupon generate instruction validation
...
 
     @ui
-    Scenario: Trying to generate a new coupons without specifying its code length
-        Given I want to generate a new coupons for this promotion
-        When I do not specify its code length
-        And I specify its amount as 4
+    Scenario: Trying to generate a new coupons without specifying their code length
+        When I want to generate new coupons for this promotion
+        And I do not specify their code length
+        When I choose the amount of 4 coupons to be generated
```
Comment (GSadee) [inline features/promotion/managing_coupons/coupon_generate_validation.feature]:
```suggestion
        And I choose the amount of 4 coupons to be generated
```

File: features/promotion/managing_coupons/generating_coupon.feature [hunk 26]
```diff
@@ -12,21 +12,21 @@ Feature: Generating a new coupons
...
+        Then I should be notified that they have been successfully generated
+        And there should be 5 coupons related to this promotion
 
     @ui
-    Scenario: Generating a new coupons without expiration date
-        Given I want to generate a new coupons for this promotion
-        When I specify its amount as 5
-        And I specify its code length as 6
+    Scenario: Generating new coupons without expiration date
+        Given I want to generate new coupons for this promotion
```
Comment (GSadee) [inline features/promotion/managing_coupons/generating_coupon.feature]:
```suggestion
        When I want to generate new coupons for this promotion
```

File: features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And there is a promotion "Christmas sale"
+        And it is coupon based promotion
+        And I am logged in as an administrator
+
+    @ui
+    Scenario: Generating new coupons with prefix
+        Given I want to generate new coupons for this promotion
```
Comment (GSadee) [inline features/promotion/managing_coupons/generating_coupon_with_prefix_and_suffix.feature]:
```suggestion
        When I want to generate new coupons for this promotion
```
and in other scenarios in this feature

File: src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionCouponsContext.php [hunk 28]
```diff
@@ -88,22 +88,38 @@ public function iWantToModifyTheCoupon(PromotionCouponInterface $coupon, Promoti
     }
 
     /**
-     * @Given /^I want to generate a new coupons for (this promotion)$/
+     * @Given /^I want to generate new coupons for (this promotion)$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionCouponsContext.php]:
```suggestion
     * @When /^I want to generate new coupons for (this promotion)$/
```

**Comment by lchrusciel** [comment]
Thank you, Mateusz! :tada:

**Pull request merged**